From 2ee511e84532e8838f2b4a31f70dd15dba3de94a Mon Sep 17 00:00:00 2001
From: Nitin Garg <nitin.garg@freescale.com>
Date: Wed, 12 Dec 2012 13:17:04 -0600
Subject: [PATCH 113/310] ENGR00236828: Add recovery ui library for iMX53 SMD

Add recovery ui library for iMX53 SMD as the selection key
is KEY_POWER instead of KEY_ENTER.

Signed-off-by: Nitin Garg <nitin.garg@freescale.com>
---
 imx53_smd/BoardConfig.mk           |  1 +
 imx53_smd/recovery/Android.mk      | 13 ++++++
 imx53_smd/recovery/recovery_ui.cpp | 92 ++++++++++++++++++++++++++++++++++++++
 3 files changed, 106 insertions(+)
 create mode 100644 imx53_smd/recovery/Android.mk
 create mode 100644 imx53_smd/recovery/recovery_ui.cpp

diff --git a/imx53_smd/BoardConfig.mk b/imx53_smd/BoardConfig.mk
index 8d86757..02a4a33 100755
--- a/imx53_smd/BoardConfig.mk
+++ b/imx53_smd/BoardConfig.mk
@@ -74,3 +74,4 @@ BOARD_KERNEL_BASE := 0x70800000
 
 TARGET_BOOTLOADER_CONFIG := mx53_smd_android_config
 
+TARGET_RECOVERY_UI_LIB := librecovery_ui_imx53_smd
diff --git a/imx53_smd/recovery/Android.mk b/imx53_smd/recovery/Android.mk
new file mode 100644
index 0000000..31523c8
--- /dev/null
+++ b/imx53_smd/recovery/Android.mk
@@ -0,0 +1,13 @@
+LOCAL_PATH := $(call my-dir)
+include $(CLEAR_VARS)
+
+LOCAL_MODULE_TAGS := eng
+LOCAL_C_INCLUDES += bootable/recovery
+LOCAL_SRC_FILES := recovery_ui.cpp
+
+# should match TARGET_RECOVERY_UI_LIB set in BoardConfig.mk
+LOCAL_MODULE := librecovery_ui_imx53_smd
+
+include $(BUILD_STATIC_LIBRARY)
+
+include $(CLEAR_VARS)
diff --git a/imx53_smd/recovery/recovery_ui.cpp b/imx53_smd/recovery/recovery_ui.cpp
new file mode 100644
index 0000000..befc181
--- /dev/null
+++ b/imx53_smd/recovery/recovery_ui.cpp
@@ -0,0 +1,92 @@
+/*
+ * Copyright (C) 2012 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* Copyright (C) 2012 Freescale Semiconductor, Inc. */
+
+#include <linux/input.h>
+
+#include "common.h"
+#include "device.h"
+#include "screen_ui.h"
+
+static const char* HEADERS[] = { "Volume up/down to move highlight;",
+                                 "power button to select.",
+                                 "",
+                                 NULL };
+
+static const char* ITEMS[] =  {"reboot system now",
+                               "apply update from ADB",
+                               "wipe data/factory reset",
+                               "wipe cache partition",
+                               NULL };
+
+class IMX53_SMD_UI : public ScreenRecoveryUI {
+  public:
+    virtual KeyAction CheckKey(int key) {
+        if (key == KEY_HOME) {
+            return TOGGLE;
+        }
+        return ENQUEUE;
+    }
+};
+
+class IMX53_SMD_Device : public Device {
+  public:
+    IMX53_SMD_Device() :
+        ui(new IMX53_SMD_UI) {
+    }
+
+    RecoveryUI* GetUI() { return ui; }
+
+    int HandleMenuKey(int key, int visible) {
+        if (visible) {
+            switch (key) {
+              case KEY_DOWN:
+              case KEY_VOLUMEDOWN:
+                return kHighlightDown;
+
+              case KEY_UP:
+              case KEY_VOLUMEUP:
+                return kHighlightUp;
+
+              case KEY_POWER:
+                return kInvokeItem;
+            }
+        }
+
+        return kNoAction;
+    }
+
+    BuiltinAction InvokeMenuItem(int menu_position) {
+        switch (menu_position) {
+          case 0: return REBOOT;
+          case 1: return APPLY_ADB_SIDELOAD;
+          case 2: return WIPE_DATA;
+          case 3: return WIPE_CACHE;
+          default: return NO_ACTION;
+        }
+    }
+
+    const char* const* GetMenuHeaders() { return HEADERS; }
+    const char* const* GetMenuItems() { return ITEMS; }
+
+  private:
+    RecoveryUI* ui;
+};
+
+Device* make_device() {
+    return new IMX53_SMD_Device();
+}
-- 
1.8.0

